Detectors/MUON/MCH/Tracking/src/TrackExtrap.cxx

```diff
@@ -0,0 +1,716 @@
+  double residue = zEnd - trackParam->getZ();
+  bool uturn = false;
+  bool trackingFailed = false;
+  bool tooManyStep = false;
+  while (TMath::Abs(residue) > SRungeKuttaMaxResidue && stepNumber <= SMaxStepNumber) {
+
+    dZ = zEnd - trackParam->getZ();
+    // step lenght assuming linear trajectory
+    step = dZ * TMath::Sqrt(1.0 + trackParam->getBendingSlope() * trackParam->getBendingSlope() +
+                            trackParam->getNonBendingSlope() * trackParam->getNonBendingSlope());
+    convertTrackParamForExtrap(trackParam, forwardBackward, v3);
+
+    do { // reduce step lenght while zEnd oversteped
+      if (stepNumber > SMaxStepNumber) {
+        LOG(WARNING) << "Too many trials: " << stepNumber;
+        tooManyStep = true;
+        break;
+      }
+      stepNumber++;
+      step = TMath::Abs(step);
+      if (!extrapOneStepRungekutta(chargeExtrap, step, v3, v3New)) {
+        trackingFailed = true;
+        break;
+      }
+      residue = zEnd - v3New[2];
+      step *= dZ / (v3New[2] - trackParam->getZ());
+    } while (residue * dZ < 0 && TMath::Abs(residue) > SRungeKuttaMaxResidue);
+
+    if (trackingFailed) {
+      break;
+    } else if (v3New[5] * v3[5] < 0) { // the track turned around
+      LOG(WARNING) << "The track turned around";
+      uturn = true;
+      break;
+    } else {
+      recoverTrackParam(v3New, chargeExtrap * forwardBackward, trackParam);
+    }
+  }
+
+  // terminate the extropolation with a straight line up to the exact "zEnd" value
+  if (trackingFailed || uturn) {
+
+    // track ends +-100 meters away in the bending direction
+    dZ = zEnd - v3[2];
+    double bendingSlope = TMath::Sign(1.e4, -sSimpleBValue * trackParam->getInverseBendingMomentum()) / dZ;
+    double pZ =
+      TMath::Abs(1. / trackParam->getInverseBendingMomentum()) / TMath::Sqrt(1.0 + bendingSlope * bendingSlope);
+    double nonBendingSlope = TMath::Sign(TMath::Abs(v3[3]) * v3[6] / pZ, trackParam->getNonBendingSlope());
+    trackParam->setNonBendingCoor(trackParam->getNonBendingCoor() + dZ * nonBendingSlope);
+    trackParam->setNonBendingSlope(nonBendingSlope);
+    trackParam->setBendingCoor(trackParam->getBendingCoor() + dZ * bendingSlope);
+    trackParam->setBendingSlope(bendingSlope);
+    trackParam->setZ(zEnd);
+
+    return false;
+
+  } else {
+
+    // track extrapolated normally
+    trackParam->setNonBendingCoor(trackParam->getNonBendingCoor() + residue * trackParam->getNonBendingSlope());
+    trackParam->setBendingCoor(trackParam->getBendingCoor() + residue * trackParam->getBendingSlope());
+    trackParam->setZ(zEnd);
+
+    return !tooManyStep;
+  }
+}
+
+//__________________________________________________________________________
+bool TrackExtrap::extrapOneStepRungekutta(double charge, double step, const double* vect, double* vout)
+{
+  /// <pre>
+  ///  ******************************************************************
+  ///  *                                                                *
```

Comment:
out of curiosity: Do we really need to code another solver for MUON tracks? Isn't this available somewhere generically? @shahor02 ?

Reply:
@sawenzel No, Muon has to use completely different track model, for the propagation in dipole field...

Reply:
Thanks.